### PR TITLE
[@lit-labs/analyzer] Add analyzer package and start of Analyzer class

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

File: packages/labs/analyzer/src/lib/analyzer.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    throw new Error('LitElement not found');
+  };
+
+  getLitElementModule = () => {
+    const files = this.program.getSourceFiles();
+    for (const file of files) {
+      if (
+        file.fileName.endsWith('/node_modules/lit-element/lit-element.d.ts')
+      ) {
+        // hopefully there's only one of these...
```
Comment (rictic) [inline packages/labs/analyzer/src/lib/analyzer.ts:81]: In many cases there will be multiple. I think we should have a set of LitElement class declarations.

Hm, or in order to be lazier, maybe we should instead look at each of the super classes' base types and spot check it to see if it's a LitElement, rather than proactively tracking down the declaration(s) in the expectation that later we'll want to know it

Comment (justinfagnani) [reply packages/labs/analyzer/src/lib/analyzer.ts:81]: Yeah, that may be nice with a cache of types that are already subclasses of LitElement

File: packages/labs/analyzer/package.json [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/lit/lit.git",
+    "directory": "packages/labs/analyzer"
+  },
+  "main": "index.js",
+  "type": "module",
+  "scripts": {
+    "build": "tsc --build",
+    "install-test-deps": "cd test-files/basic-elements && npm ci",
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/analyzer/package.json:20]: it may be enough to express "before running tests, if any of the package.jsons in test-files have changed, then install test deps"

**Review by rictic — APPROVED** [review]

File: packages/labs/analyzer/src/test/analyzer_test.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import ts from 'typescript';
+import * as path from 'path';
+
+import {Analyzer} from '../lib/analyzer.js';
+
+test('Basic elements - isLitElement', () => {
+  const basicElementsPackagePath = new URL(
+    '../test-files/basic-elements',
+    import.meta.url
+  ).pathname;
```
Comment (aomarks) [inline packages/labs/analyzer/src/test/analyzer_test.ts:18]: Won't this always be a `/` delimited path? So it won't work on Windows. I think you want:

```ts
import * as pathlib from 'path';
import {fileURLToPath} from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = pathlib.dirname(__filename);
const basicElementsPackagePath = pathlib.resolve(__dirname, '..', 'test-files', 'basic-elements');
```

File: packages/labs/analyzer/test-files/basic-elements/src/element-a.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import {LitElement, html, css} from 'lit';
```
Comment (aomarks) [inline packages/labs/analyzer/test-files/basic-elements/src/element-a.ts:1]: Even test files should probably have copyright statement?

File: packages/labs/analyzer/src/test/analyzer_test.ts (hunk 3 above)
Comment (aomarks) [inline packages/labs/analyzer/src/test/analyzer_test.ts:18]: Speaking of Windows support, I think it would be a good idea to start testing on Windows right away, since it's easier to handle Windows edge cases incrementally instead of in a big pass afterwards. It's pretty to set up Windows CI with GitHub Actions, see https://github.com/lit/wireit/blob/97b7a9f3932597a8c91d3a4d406fcdd7251d234d/.github/workflows/tests.yml#L9

(Another PR fine, I just recommend doing it sooner rather than later)

File: packages/labs/analyzer/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (aomarks) [reply packages/labs/analyzer/package.json:20]: Maybe a path `devDependency` dependency would work? Not sure if it matters that that could cause some dependencies to be installed/deduped into `analyzer/node_modules` instead of `basic-elements/node_modules`?

File: packages/labs/analyzer/src/test/analyzer_test.ts (hunk 3 above)
Comment (aomarks) [reply packages/labs/analyzer/src/test/analyzer_test.ts:18]: I guess this would work too, a bit simpler:

```ts
import {fileURLToPath} from 'url';

fileURLToPath(new URL(
  '../test-files/basic-elements',
  import.meta.url
).pathname));
```

File: packages/labs/analyzer/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,31 @@
+{
+  "compilerOptions": {
+    "incremental": true,
+    "composite": true,
+    "tsBuildInfoFile": "./lib/.tsbuildinfo",
+    "target": "es2020",
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/analyzer/tsconfig.json:6]: Ah yeah, we might also want to use real private fields in the analyzer

[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/labs/analyzer/tsconfig.json:6]: Yeah, I think Node 14 is reasonable for now. It'll go out of maintenance 2023-04-23.

Class fields, including private, are es2022 though. tsc's downlevel of them isn't great either, IMO, so I'd rather not.

**Review by aomarks — APPROVED** [review]

File: packages/labs/analyzer/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/labs/analyzer/package.json]: that did work. It did also hoist `@lit/...` into the analyzer's node_modules, but I think that's ok.

File: packages/labs/analyzer/src/lib/analyzer.ts [hunk 6]
```diff
@@ -81,10 +81,7 @@ export class Analyzer {
   };
 
   private _isLitElementModule = (file: ts.SourceFile) => {
-    return (
-      file.fileName.startsWith(this.packageRoot) &&
-      file.fileName.endsWith('/node_modules/lit-element/lit-element.d.ts')
-    );
+    return file.fileName.endsWith('/node_modules/lit-element/lit-element.d.ts');
```
Comment (rictic) [inline packages/labs/analyzer/src/lib/analyzer.ts:84]: just curious, did something come up that prompted this change?

Comment (justinfagnani) [reply packages/labs/analyzer/src/lib/analyzer.ts:84]: Yeah: `lit-element` got hoisted out of the test-project when I made the test project a devDep of analyzer. We probably want to support hoisted installs for monorepos and things like pnpm anyway.
